[AV2-841] Feature/av2 841/update program switcher component to experience switcher

sasangachathumal opened the pull request:
**Story/Ticket Reference ID from JIRA:**
https://intersective.atlassian.net/browse/AV2-841

### Additional Comments

**This PR only have the code changes. so unit test in PR is falling. And I have create separate PR and branch for Unit testing because changes getting bigger so it's hard to code review track.** 

this is the other PR I created for Unit testing and i'm still working on it. https://github.com/intersective/app-v2/pull/1240

Sunilsbcloud committed:
uk bring-up

Sunilsbcloud committed:
uk bring-up

Sunilsbcloud committed:
uk bring-up

Sunilsbcloud committed:
uk bring-up

Sunilsbcloud committed:
uk bring-up

Sunilsbcloud committed:
uk bring-up

Sunilsbcloud committed:
uk bring-up

Sunilsbcloud committed:
uk bring-up

Sunilsbcloud committed:
Merge branch 'develop' into feature/uk_release_fix

Sunilsbcloud committed:
uk bring-up

Sunilsbcloud committed:
uk bring-up

shawnm0705 committed:
Merge pull request #1211 from intersective/feature/uk_release_fix

Feature/uk release fix

Sunilsbcloud committed:
Merge pull request #1213 from intersective/develop

sync

trtshen committed:
[AV2-841] cherry-pick some changes.

sasangachathumal committed:
[AV2-929]
remove code, making cursor empty every time there are no messages.

sasangachathumal committed:
Merge pull request #1218 from intersective/feature/AV2-929/chat-messages-duplicate-when-scrolling-up

[AV2-929] Chat messages duplicate visually when scrolling up to see older messages.

sasangachathumal committed:
Merge pull request #1219 from intersective/develop

develop to trunk

trtshen committed:
[CORE-4599] "review submitted"

Sunilsbcloud committed:
fix workflow error

Sunilsbcloud committed:
Merge pull request #1221 from intersective/feature/cicd-fix

fix workflow error

Sunilsbcloud committed:
Merge pull request #1222 from intersective/develop

sync

trtshen committed:
Update assessment.component.spec.ts

trtshen committed:
[CORE-4599]

trtshen committed:
Update assessment.component.ts

trtshen committed:
Update assessment.component.ts

trtshen committed:
Update assessment.component.ts

shawnm0705 committed:
Merge pull request #1220 from intersective/bugfix/CORE-4599/infinite

[CORE-4599] Bugfix/infinite-loop

trtshen committed:
Merge branch 'trunk' into bugfix/CORE-4599/infinite

shawnm0705 committed:
Merge pull request #1223 from intersective/bugfix/CORE-4599/infinite

[CORE-4599] Bugfix/infinite

trtshen committed:
Update app.component.ts

trtshen committed:
Update app.component.ts

trtshen committed:
Merge pull request #1224 from intersective/hotfix/clearing-cached-logo

allow null config

trtshen committed:
Update app.component.ts

trtshen committed:
Update app.component.ts

trtshen committed:
Update app.component.ts

trtshen committed:
Update app.component.ts

sasangachathumal committed:
Merge pull request #1225 from intersective/hotfix/clearing-cached-logo

Update app.component.ts

shawnm0705 committed:
Merge pull request #1226 from intersective/config-fixes-from-trunk

Config fixes from trunk

sasangachathumal committed:
[AV2-841]
Updated post method params to json object.
Update every service calling post request.

sasangachathumal committed:
[AV2-841]
Modify the routes in tabs.
added loading skeleton to switcher page.
update switcher component to get expreances.
update seitcher service to call each stack.
added new method to format each stack response.

shawnm0705 committed:
[AV2-931]

shawnm0705 committed:
[AV2-931]

shawnm0705 committed:
package lock

shawnm0705 committed:
[AV2-931]

shawnm0705 committed:
[AV2-931]

shawnm0705 committed:
Merge pull request #1228 from intersective/feature/av2-931/disable-intercom,-piwik,-newrelic

[AV2-931]disable intercom, piwik, newrelic

shawnm0705 committed:
Merge pull request #1229 from intersective/develop

develop to trunk

Sunilsbcloud committed:
newrelic-disable-env

shawnm0705 committed:
Merge pull request #1230 from intersective/feature/newrelic-disable-env

newrelic-disable-env

sasangachathumal committed:
[AV2-841]
modify the clickable-item component to work in switcher page.
added experiense status element to switcher page.
added new image for skeleton loading in switcher page.
created new scss variables for experiense status colors.
added styles for experiense status elements.

shawnm0705 committed:
[AV2-931]

shawnm0705 committed:
[AV2-931]

shawnm0705 committed:
Merge pull request #1231 from intersective/feature/av2-931/disable-intercom,-piwik,-newrelic

[AV2-931]

shawnm0705 committed:
Merge pull request #1233 from intersective/develop

develop to trunk

trtshen committed:
[CORE-4694]

trtshen committed:
Update utils.service.ts

sasangachathumal committed:
Merge branch 'global-appv2'

shawnm0705 committed:
[AV2-932]

shawnm0705 committed:
Merge pull request #1234 from intersective/feature/av2-932/add-submission-id-in-the-assessment-route

[AV2-932]add submission id in the assessment route

shawnm0705 committed:
Merge pull request #1235 from intersective/develop

develop to trunk

shawnm0705 committed:
[AV2-932]

shawnm0705 committed:
[AV2-932]

trtshen committed:
[CORE-4694] handle 3 colors now

sasangachathumal committed:
[AV2-841]
Moved pusher and apollo init to shared service.
pusher and apollo init call in switcher service and app component.
removed apiurl from pusher module and service.
update switch resolver to save stack in local storage.
added handler to each auth api call to add stack obj to response.
added code to sort by date.
update html to show experience status.
add code to save apikey, stack config, progroms local storage.

shawnm0705 committed:
Merge pull request #1236 from intersective/feature/av2-932/add-submission-id-in-the-assessment-route

[AV2-932]add submission id in the assessment route

trtshen committed:
[CORE-4694]

trtshen committed:
Update utils.service.spec.ts

trtshen committed:
[CORE-4694]

shawnm0705 committed:
Merge pull request #1237 from intersective/develop

develop to trunk

trtshen committed:
[CORE-4694]

trtshen committed:
[CORE-4694]

trtshen committed:
Update storage.service.spec.ts

trtshen committed:
Update storage.service.ts

trtshen committed:
Update auth-reset-password.component.spec.ts

trtshen committed:
[CORE-4694]

trtshen committed:
[CORE-4694]

sasangachathumal committed:
[AV2-841]
Fixed unit test for POST request method update in services.
fixed unit test for shared service update.
update pusher service unit test.

shawnm0705 committed:
Merge pull request #1239 from intersective/hotfix/CORE-4694/primary-2ndary-colors

[CORE-4694] Hotfix/primary 2ndary colors

sasangachathumal commented:
This PR only have the code changes. so unit test in PR is falling. And I have create separate PR and branch for Unit testing because changes getting bigger so it's hard to code review track.

Can you guys review this PR for me.

this is the other PR I created for Unit testing and i'm still working on it. https://github.com/intersective/app-v2/pull/1240

trtshen commented on src/app/switcher/switcher.service.ts:
```diff
@@ -129,6 +188,9 @@ export class SwitcherService {
 
   switchProgram(programObj: ProgramObj): Observable<any> {
+    // initialise Pusher and apollo here if there stack info in storage
+    this.sharedService.initPusherApollo();
```
Hi @sasangachathumal I prefer to call this "initWebServices" so we can reuse that for different services in the future.

sasangachathumal replied:
sure. will change the naming.

trtshen commented on src/app/switcher/switcher.service.ts:
```diff
@@ -76,23 +80,78 @@ export class SwitcherService {
+    const expreancesList = [];
+    apiResults.forEach(result => {
+      const data = result.data;
+      data.Timelines.map(
```
@sasangachathumal I'm not sure if this possible, do we need to cover condition where the  `Timelines` object from data is null or undefined?

sasangachathumal replied:
i will check, better to have something like that. i will check and add.

shawnm0705 commented on src/app/switcher/switcher-program/switcher-program.component.ts:
```diff
@@ -42,50 +43,36 @@ export class SwitcherProgramComponent extends RouterEnter implements AfterConten
+    this.switcherService.getExperience(this.stacks).subscribe(
+      experiences => {
+        this.isExperiencesLoading = false;
+        this.programs = experiences;
```
let's name it consistently among the codebase, either "program" everywhere, or "experience" everywhere.

sasangachathumal committed:
[AV2-841] update the method naming.

shawnm0705 commented on src/app/switcher/switcher.service.ts:
```diff
@@ -76,23 +80,78 @@ export class SwitcherService {
+                theme_color: 'var(--ion-color-primary)'
+              };
+            } else {
+              timeline.Program.config.theme_color = 'var(--ion-color-primary)';
```
I think @trtshen has latest code changes related to the color. @sasangachathumal you may want to implement them here as well

trtshen replied:
@sasangachathumal  https://github.com/intersective/app-v2/tree/hotfix/CORE-4694/primary-2ndary-colors

trtshen replied:
@sasangachathumal i've created a branch  to make your work easier, it has all the color changes code i did. you can merge this changes back into your current work later.

sasangachathumal replied:
sure. thank you @trtshen .

shawnm0705 commented on src/app/switcher/switcher.service.ts:
```diff
@@ -76,23 +80,78 @@ export class SwitcherService {
+            }
+            timeline.Project.lead_image = `${cdn}${imagewidth}/${imageId}`;
+          }
+          expreancesList.push({
```
typo

sasangachathumal committed:
[AV2-841] updated to "program" everywhere.

sasangachathumal committed:
[AV2-841]
simplified the normalise method of auth result.
added new code to it check is timeline is array.
added new code to redirect user to login if no programs found.

sasangachathumal committed:
Merge /hotfix/CORE-4694/primary-2ndary-colors

sasangachathumal committed:
[AV2-841]
remove apikey from auth.json request headers.

sasangachathumal committed:
Merge feature/AV2-841/experience-switcher

sasangachathumal committed:
[AV2-841]
added unit test to app component.
added unit test for shared service for init web services.
modifed storage service unit tests.
update intercepter unit test to chek auth.json.
added unit test for switcher service to cover new changes.
fixed some code bugs in switcher service.
added unit test for switcher component for new changes.
update mock service with new changes.

sasangachathumal committed:
Merge pull request #1240 from intersective/feature/AV2-841/unit-test

[AV2-841] Unit test for Experience switcher changes.

trtshen commented on src/app/shared/request/request.service.ts:
```diff
@@ -150,29 +158,31 @@ export class RequestService {
-  post(endPoint: string = '', data, httpOptions?: RequestOptions, isLoginAPI?: boolean): Observable<any> {
-    if (!httpOptions) {
-      httpOptions = {};
+  post(params: POSTParams): Observable<any> {
```
@sasangachathumal  👍  i like how this got cleaned up!

sasangachathumal replied:
Thank you @trtshen 👍. Need to do same refactoring for other functions like get in the future.
